nfs/src/rpc_task.cpp

```diff
@@ -304,6 +501,294 @@ void rpc_task::run_lookup()
+                    src_off += res;
+
+                    // Not able to read full from fd, and retry not allowed.
+                    if (len && !retry_set){
+                        assert(0);
+                    }
+                }
+            } else {
+                // buffer copy case.
+                memcpy(chunk.get_buffer(), buf, chunk.length);    
+            }
+
+        } else {
+                // Need to issue read.
+        }
+
+        if (error == 0) {
+            membuf->set_dirty();
+            membuf->set_uptodate();
+        } else {
+            membuf->clear_locked();
+            break;
+        }
+
+        membuf->clear_locked();
+
+        if (buf != nullptr) {
+            buf += chunk.length;
+        }
+
+        rem_len -= chunk.length;
+    }
+    
+    if (error < 0) {
+        AZLogDebug("Error is set, releasing chunkMap ({}, {})", offset, length);
+        for (auto &chunk : chunkvec)
+        {
+            auto membuf = chunk.get_membuf();
+
+            // Clear the membuf inuse flag.
+            membuf->clear_inuse();
+        }
+        file_inode->filecache_handle->release(offset, length);
+        return std::vector<bytes_chunk>();
+    }
+
+    assert (rem_len == 0);
+    
+    return chunkvec;
+}
+
+void rpc_task::run_write()
+{
+    fuse_ino_t file_ino = rpc_api.write_task.get_ino();
+    auto nfs_inode = get_client()->get_nfs_inode_from_ino(file_ino);
+    size_t length = rpc_api.write_task.get_size();
+    struct fuse_bufvec* bufv = rpc_api.write_task.get_buffer_vector();
+    off_t offset = rpc_api.write_task.get_offset();
+    auto client = get_client();
+    std::vector<bytes_chunk> chunkvec;
+
+    int error_code = nfs_inode->get_write_error();
+    if (error_code == 0)
+    {
+        chunkvec = copy_to_cache(get_client(), file_ino, bufv, offset, length, error_code);
+    }
+
+    if (error_code < 0) {
+        AZLogDebug("copy_to_cache return error={}", error_code);
+
+        assert(chunkvec.size() == 0);
+        reply_error(error_code);
+        return;
+    }
+
+    assert(chunkvec.size() > 0);
+
+    for (auto &chunk : chunkvec) {
+        WRITE3args args;
+        bool rpc_retry = false;
+
+        // Get the membuf which need to flush.
+        auto membuf = chunk.get_membuf();
+        membuf->set_locked();
+
+        assert(membuf->is_dirty());
+        assert(membuf->is_uptodate());
+
+        // Fill the Write RPC arguments.
+        args.file = client->get_nfs_inode_from_ino(file_ino)->get_fh();
+        args.offset = membuf->offset;
+        args.count  = membuf->length;
+        args.stable = FILE_SYNC;
+        args.data.data_len = membuf->length;
+        args.data.data_val = (char *) membuf->buffer;
+
+        // Allocate new rpc task.
+        auto flush_task = client->get_rpc_task_helper()->alloc_rpc_task();
+        flush_task->optype = FUSE_FLUSH;
+
+        struct write_flush_context *cb_context = new write_flush_context(chunk, flush_task, file_ino);
```

Comment:
why do you need to create an rpc task here? 
see related comment in run_flush()

Reply:
As discussed to put back pressure on incoming writes, created new rpc.